src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1551,21 +1534,29 @@ impl Disambiguator {
                 "prim" | "primitive" => Primitive,
                 _ => return Err((format!("unknown disambiguator `{}`", prefix), 0..idx)),
             };
-            Ok(Some((d, &rest[1..])))
+            Ok(Some((d, &rest[1..], rest[1..].trim().to_owned())))
         } else {
             let suffixes = [
                 ("!()", DefKind::Macro(MacroKind::Bang)),
+                ("!{}", DefKind::Macro(MacroKind::Bang)),
+                ("![]", DefKind::Macro(MacroKind::Bang)),
                 ("()", DefKind::Fn),
                 ("!", DefKind::Macro(MacroKind::Bang)),
             ];
             for (suffix, kind) in suffixes {
-                if let Some(link) = link.strip_suffix(suffix) {
+                if let Some(path_str) = link.strip_suffix(suffix) {
                     // Avoid turning `!` or `()` into an empty string
-                    if !link.is_empty() {
-                        return Ok(Some((Kind(kind), link)));
+                    if !path_str.is_empty() {
+                        return Ok(Some((Kind(kind), path_str, link.trim().to_owned())));
                     }
                 }
             }
+            // Parse `macro!\[\]` (backslashes are needed in link labels)
+            if let Some(path_str) = link.strip_suffix(r"!\[\]") {
```

Comment:
This seems like a misfeature (it also supports, for example, `[m](m!\[\])` and `[m]: m!\[\]`). I don't think we should adapt the markdown syntax, people will find it surprising that `![]` isn't supported either way and there are plenty of other workarounds.

If you remove this, please also change the return type from String to &str.